drv/i2c-devices/src/pca9956b.rs

```diff
@@ -0,0 +1,311 @@
+/// The other u32 fields are masks of which LED output the error was observed.
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq)]
+pub struct Pca9956BErrorState {
+    pub overtemp: bool,
+    pub errors: [LedErr; NUM_LEDS],
+}
+
+/// Auto-increment flag is Bit 7 of the control register. Bits 6..0 are address.
+const CTRL_AUTO_INCR_MASK: u8 = 1 << 7;
+/// The MODE2 OVERTEMP bit indicates if an overtempature condition has occurred
+const MODE2_OVERTEMP_MASK: u8 = 1 << 7;
+/// The MODE2 ERROR bit indicates if any error conditions are in EFLAGn
+const MODE2_ERROR_MASK: u8 = 1 << 6;
+/// The MODE2 CLRERR bit clears all error conditions in EFLAGn
+const MODE2_CLRERR_MASK: u8 = 1 << 4;
+/// The MODE2 reserved bits have a defined pattern (0b101) and are read only.
+/// They will be used to validate a PCA9956B with humility
+const MODE2_RSVD_MASK: u8 = 0x7;
+const MODE2_RSVD: u8 = 0x5;
+
+pub struct Pca9956B {
+    device: I2cDevice,
+}
+
+pub const NUM_LEDS: usize = 24;
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum Error {
+    /// The low-level I2C communication returned an error
+    I2cError(ResponseCode),
+
+    /// The LED index is too large
+    InvalidLED(u8),
+
+    /// Write buffer too large
+    WriteBufferTooLarge(usize),
+
+    /// Register is outside a valid auto-increment range
+    InvalidAutoIncReg(Register),
+}
+
+impl From<ResponseCode> for Error {
+    fn from(err: ResponseCode) -> Self {
+        Error::I2cError(err)
+    }
+}
+
+impl Pca9956B {
+    pub fn new(device: &I2cDevice) -> Self {
+        Self { device: *device }
+    }
+
+    /// Read a single Register
+    fn read_reg(&self, reg: Register) -> Result<u8, Error> {
+        self.device
+            .read_reg::<u8, u8>(reg as u8)
+            .map_err(Error::I2cError)
+    }
+
+    /// Read a number of Registers into `buf`
+    /// Note that CTRL_AUTO_INCR_MASK is set here to enable the device's
+    /// auto-increment feature, which has varying behaviors and limitations that
+    /// must be accounted for.
+    #[allow(dead_code)]
+    fn read_buffer(&self, reg: Register, buf: &mut [u8]) -> Result<(), Error> {
+        self.device
+            .read_reg_into((reg as u8) | CTRL_AUTO_INCR_MASK, buf)
+            .map_err(Error::I2cError)?;
+
+        Ok(())
+    }
+
+    /// Write a single Register
+    fn write_reg(&self, reg: Register, val: u8) -> Result<(), Error> {
+        let buffer = [reg as u8, val];
+        self.device.write(&buffer).map_err(Error::I2cError)
+    }
+
+    /// Write a number of Registers into `buf`
+    /// Note that CTRL_AUTO_INCR_MASK is set here to enable the device's
+    /// auto-increment feature, which has varying behaviors and limitations that
+    /// must be accounted for.
+    fn write_buffer(&self, reg: Register, buf: &[u8]) -> Result<(), Error> {
+        if buf.len() > MAX_BUF_SIZE {
+            return Err(Error::WriteBufferTooLarge(buf.len()));
+        } else if reg > MAX_AUTO_INC_REG {
```

Comment:
I was a bit conflicted on how opinionated to be with these error conditions because while there is a small subset of registers where auto-increment doesn't work, if you begin reading/writing _outside_ that range, the PCA9956B rolls its internal counter over. So, for instance, if you begin writing to `0x3E`, the next write will occur at `0x0` (in the default auto-increment configuration). However if you try to begin an auto-incrementing transaction at `0x3F` the behavior is undefined. So my decision here was to prevent someone from directly doing something undefined, while leaving open the opportunity to leverage the internal address rollover if they chose to.